packages/language-service/ivy/utils.ts

```diff
@@ -0,0 +1,196 @@
+  }
+
+  return undefined;
+}
+
+/**
+ * Retrieves the `ts.ClassDeclaration` at a location along with its template nodes.
+ */
+function getInlineTemplateInfoAtPosition(
+    fileName: string, position: number, program: ts.Program,
+    templateTypeChecker: TemplateTypeChecker):
+    {template: TmplAstNode[], component: ts.ClassDeclaration}|undefined {
+  const sourceFile = program.getSourceFile(fileName);
+  if (!sourceFile) {
+    return undefined;
+  }
+
+  for (const statement of sourceFile.statements) {
+    if (!ts.isClassDeclaration(statement) || position < statement.pos || position > statement.end) {
+      continue;
+    }
+
+    const template = templateTypeChecker.getTemplate(statement);
+    if (template === null) {
+      return undefined;
+    }
+
+    return {template, component: statement};
+  }
+
+  return undefined;
+}
+
+/**
+ * Given an attribute name and the element or template the attribute appears on, determines which
+ * directives match because the attribute is present. That is, we find which directives are applied
+ * because of this attribute by elimination: compare the directive matches with the attribute
+ * present against the directive matches without it. The difference would be the directives which
+ * match because the attribute is present.
+ *
+ * @param attribute The attribute name to use for directive matching
+ * @param hostNode The element or template node that the attribute is on
+ * @param directives The list of directives to match against
+ */
+export function getDirectiveMatchesForAttribute(
+    attribute: string, hostNode: TmplAstTemplate|TmplAstElement, directives: DirectiveSymbol[]) {
+  let attributes: Array<TmplAstTextAttribute|TmplAstBoundAttribute> =
+      [...hostNode.attributes, ...hostNode.inputs];
+  if (hostNode instanceof TmplAstTemplate) {
+    attributes = [...attributes, ...hostNode.templateAttrs];
+  }
+  const toAttributeString = (a: TmplAstTextAttribute|TmplAstBoundAttribute) =>
+      `[${a.name}=${a.valueSpan?.toString() ?? ''}]`;
+  const attrs = attributes.map(a => toAttributeString(a));
+  const attrsOmit = attributes.map(a => a.name === attribute ? '' : toAttributeString(a));
+
+  const hostNodeName = hostNode instanceof TmplAstTemplate ? hostNode.tagName : hostNode.name;
+  const directivesWithAttribute = getDirectiveMatches(directives, hostNodeName + attrs.join(''));
+  const directivesWithoutAttribute =
+      getDirectiveMatches(directives, hostNodeName + attrsOmit.join(''));
+  return directivesWithAttribute.filter(d => !directivesWithoutAttribute.some(s => s === d));
+}
+
+/**
+ * Returns a new `ts.SymbolDisplayPart` array which has the alias imports from the tcb filtered
+ * out, i.e. `i0.NgForOf`.
+ */
+export function filterAliasImports(displayParts: ts.SymbolDisplayPart[]) {
+  return displayParts.filter((part, i) => {
+    const tcbAliasImportRegex = /i\d+/;
+    const previousPart = displayParts![i - 1];
+    const nextPart = displayParts![i + 1];
```

Comment:
```suggestion
    const previousPart = displayParts[i - 1];
    const nextPart = displayParts[i + 1];
```

I don't think the ! is needed